Add an `isEmptyRender()` method to both `ShallowWrapper` and `ReactWrapper`

milesj opened the pull request:
# Why

Added an `isNull()` method to both `ShallowWrapper` and `ReactWrapper`,
making it easier in tests to detect when a component renders nothing;
either a null or undefined value.

I chose the name `isNull()` since `isEmpty()` was used for other situations.
It's not as descriptive as it could be, so open to suggestion.

# Tested

Unit tests!

sjdemartini commented on src/ReactWrapper.js:
```diff
@@ -288,6 +288,15 @@ export default class ReactWrapper {
+   *
+   * @returns {boolean}
+   */
+  isNull() {
```
I feel like this name too specifically implies that the component did `return null` in its `render` method, even though `return undefined` (either explicitly or implicitly) should also be fine.

ljharb commented on src/ReactWrapper.js:
```diff
@@ -288,6 +288,15 @@ export default class ReactWrapper {
+   * @returns {boolean}
+   */
+  isNull() {
+    return this.single(n => findDOMNode(n) === null);
```
`return this.type() === null`?

milesj replied:
That doesn't actually work. It also seems like SFC causes issues too. Trying to figure that out.

ljharb commented on src/ShallowWrapper.js:
```diff
@@ -254,6 +254,15 @@ export default class ShallowWrapper {
+   * @returns {boolean}
+   */
+  isNull() {
+    return (this.node === null);
```
please remove the unnecessary parens.

if this is the canonical location for "is null", we should call into `isNull` in a number of other places, like `.type()` for example.

milesj commented:
Seems like a lot changed internally during the React 15 upgrade. I've not been able to find an implementation that works across all versions. Using `this.type()` doesn't work for all use cases, nor does `findDOMNode()`, so I had to use a conditional.

lelandrichardson commented:
I think we need to duplicate all of these tests for `undefined` and `false`, which are also things that are "rendering nothing" that we would probably want this function to behave identically for. Looking at the code, I don't think this will pass right now.

The other question for me is the name.  `isNull()` is pretty directly indicating that we are testing for `null`, which might not always be the case (other things can return `ReactNodeTypes.EMPTY`).

That said, I'm not really sure what the right name is. Maybe it should be `isEmpty()` or `isNothing()`?

milesj commented:
Perhaps something like `hasRenderedEmpty()` or `hasRenderedNothing()`?

milesj commented:
Ok, I changed it to `isEmptyRender()`, to be a bit more explicit. I also added `false` based tests and tried to add `undefined`, but supposedly you can't, as React errors with "Invariant Violation: Unexpected node: undefined" or "Error: Invariant Violation: Foo.render(): A valid ReactComponent must be returned. You may have returned undefined, an array or some other invalid object.".

lelandrichardson commented:
Hey @milesj , you're right. My bad - I guess undefined is not valid here.

We also want to ensure that this returns false in the proper cases though.

`false` => `true`
`null` => `true`
`true` => `false`
`0` => `false`
`<div />` => `false`
`<Foo />` => `false`


At the moment, `isEmptyRender() { return true; }` would pass all of these tests :)

I think at that point this will be ready to merge. Not super psyched about the name, but can't think of anything better at the moment.... so at least it's explicit.

milesj commented:
I was a bit surprised about the undefined thing as well. I could of sworn it allowed it at some point.

I'll look into adding some truthy tests.

ljharb commented:
Don't forget about `NaN` and the empty string, just to make sure the other falsy values are covered (both should return `false` i believe)

Aweary commented:
@milesj any updates on this? 😄

milesj commented:
Oops, thanks for reminding me. Been pretty swamped at work. Will try to get it this week.

Aweary commented:
No worries, don't feel pressured/rushed! Thanks for the contribution, looking forward to seeing how this turns out 👍

milesj commented:
Added a ton of return value checks.

milesj commented:
@lelandrichardson I added tests that check null/false, as well as React components/elements. I originally had it testing every possible falsey value, but then it just barfed all over the output, as seen by the screenshot. I don't really see the benefit of testing all the falsey values as React doesn't allow anything besides null/false.

![screen shot 2016-05-19 at 12 37 07 pm](https://cloud.githubusercontent.com/assets/143744/15408857/181ab442-1dc7-11e6-8428-554db783e25d.png)

Aweary commented:
@milesj I think we should only be testing valid  "empty" return values 👍

ljharb commented:
What about `NaN`, and the empty string? Does React complain about those too?

milesj commented:
I tested a ton of values, all of which are: null, false, undefined, NaN, array, empty array, object, empty object, zero, string, empty string, React element/component, DOM element, etc.

All of them threw those warnings excluding null/false/React.

milesj commented:
Screenshot I took.
![screen shot 2016-05-19 at 2 22 17 pm](https://cloud.githubusercontent.com/assets/143744/15410158/21d3c2f2-1dcd-11e6-8cdf-10a1459fe4bd.png)

Aweary commented:
LGTM

milesj commented:
@lelandrichardson Anything else you'd like done on this? Besides hitting the "Update Branch" button :P

Aweary commented:
@milesj can you squash this to get read of all those merge commits? I think we can merge this after that 👍

milesj commented:
Wouldn't Github's new merge squash feature suffice here?

I've tried rebasing the history a bit, but it's turned out troublesome since the commits happened at different points in history. I'll mess around some more.

Aweary commented:
Does Github do that now? I didn't realize. I'll merge.

ljharb commented:
(merge-squash is fine when the PR author has issues with rebasing, but in general we do strongly prefer rebased and non-squash-merged commits. github's squash-merge breaks the history between the PR and master)

milesj commented:
Yeah, GitHub's automatic "Update Branch" is useful, but annoying history wise. Will probably do it manually going forward.
